service/worker/versioning/workflow.go

```diff
@@ -0,0 +1,172 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package versioning
+
+import (
+	"time"
+
+	enumspb "go.temporal.io/api/enums/v1"
+	sdkclient "go.temporal.io/sdk/client"
+	sdklog "go.temporal.io/sdk/log"
+	"go.temporal.io/sdk/temporal"
+	"go.temporal.io/sdk/workflow"
+)
+
+type (
+	TaskQueue struct {
+		Name          string
+		TaskQueueType enumspb.TaskQueueType
+	}
+	Deployment struct {
+		Namespace      string
+		DeploymentName string
+		TaskQueue      []*TaskQueue // All the task queues associated with this buildID/deployment
+		BuildID        string
+	}
+	DeploymentName struct {
+		Namespace      string
+		Name           string
+		CurrentBuildID string // denotes the current "default" build-ID of a deploymentName
+	}
+
+	// DeploymentWorkflowRunner holds the local state while running a deployment workflow
+	DeploymentWorkflowRunner struct {
+		deployment Deployment
+		ctx        workflow.Context
+		a          *activities
+		logger     sdklog.Logger
+		metrics    sdkclient.MetricsHandler
+	}
+	// DeploymentWorkflowRunner holds the local state while running a deployment name workflow
+	DeploymentNameWorkflowRunner struct {
+		deploymentName DeploymentName
+		ctx            workflow.Context
+		a              *activities
+		logger         sdklog.Logger
+		metrics        sdkclient.MetricsHandler
+	}
+	// DeploymentBuildIDArgs holds the arguments used while calling activities associated with a deployment
+	DeploymentBuildIDArgs struct {
+		deployment Deployment
+		buildID    string // used for verifying/updating the buildID of all the task queues in a deployment
+	}
+)
+
+var (
+	defaultActivityOptions = workflow.ActivityOptions{
+		ScheduleToCloseTimeout: 1 * time.Hour,
+		StartToCloseTimeout:    30 * time.Second,
+		RetryPolicy: &temporal.RetryPolicy{
+			InitialInterval: 1 * time.Second,
+			MaximumInterval: 60 * time.Second,
+		},
+	}
+)
+
+const (
+	UpdateDeploymentSignalChannelName = "update_deployment"
+	updateBuildIDSignalChannelName    = "update_task_queue_build_id"
+)
+
+func DeploymentWorkflow(ctx workflow.Context, deploymentWorkflowArgs Deployment) error {
+	deploymentWorkflowRunner := &DeploymentWorkflowRunner{
+		deployment: deploymentWorkflowArgs,
+		ctx:        ctx,
+		logger:     sdklog.With(workflow.GetLogger(ctx), "wf-namespace", deploymentWorkflowArgs.Namespace),
+		metrics:    workflow.GetMetricsHandler(ctx).WithTags(map[string]string{"namespace": deploymentWorkflowArgs.Namespace}),
+	}
+	// TQ should be set to nil here since signal handler updates it
+	deploymentWorkflowRunner.deployment.TaskQueue = nil
+	return deploymentWorkflowRunner.run()
+}
+
+func (d *DeploymentWorkflowRunner) run() error {
+
+	// Set up Query Handlers here:
+
+	// Fetch signal channels
+	updateDeploymentSignalChannel := workflow.GetSignalChannel(d.ctx, UpdateDeploymentSignalChannelName)
+	updateBuildIDSignalChannel := workflow.GetSignalChannel(d.ctx, updateBuildIDSignalChannelName)
+	var signalCount int
```

Comment:
don't need this

Reply:
I wanted to keep it as a means of debugging if there was a time we would need to verify the number of signals received/processed == number of signals sent lol - I can remove it for now